Chemist Heirloom

DrDuckedGoose opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

New chemist heirloom, the 'Hard locked bottle of-'
This adds a new chemist heirloom that takes the form of an unopenable bottle filled with a random reagent. This reagent is inaccessible by all means. The only point where the reagent can be accessed is at the end of the round, where the bottle unlocks when the shuttle reaches centcomm. 

Chemist heirloom, 100u sealed bottle of random chem. Can't open.

## Why It's Good For The Game

Funny bottle is particularly funny, perhaps even humorous. 
Anything is probably better than the current placeholder, chemistry guide.

## Testing Photographs and Procedure
<details>
<summary>Screenshots&Videos</summary>
Spawn as chemist with heirloom trait, or spawn item in through debug menu. Attempt to extract chemicals in any form. End round.

Coder sprites.

![image](https://user-images.githubusercontent.com/40559528/168785482-082be33f-07a3-4992-a1ae-e151e996de36.png)
![image](https://user-images.githubusercontent.com/40559528/168785532-0bb4a9e4-50ef-4c0a-9523-98cfbbb81395.png)
![image](https://user-images.githubusercontent.com/40559528/168785839-8fcd4664-91df-4ae6-ab25-243faece79b0.png)
![image](https://user-images.githubusercontent.com/40559528/168786831-af00859d-ddec-41bb-a98d-dc73bff2ed9f.png)
![image](https://user-images.githubusercontent.com/40559528/168788560-981a3ed7-44e5-45c2-bcae-3937acee014a.png)

</details>

## Changelog
:cl:
add: chem_heirloom.dm
tweak: change chemist's heirloom to hard locked bottle
tweak: negative.dm calls chem_heirloom edgecase proc for naming.
/:cl:

![image](https://user-images.githubusercontent.com/40559528/168790101-2e6bbf11-9a57-41e4-afa7-61fd6da8c5ae.png)

DrDuckedGoose committed:
Initial

DrDuckedGoose committed:
Batch 1

DrDuckedGoose committed:
Update chem_heirloom.dm

EvilDragonfiend commented on code/modules/reagents/reagent_containers/chem_heirloom.dm:
```diff
@@ -0,0 +1,46 @@
+	name ="[name] [initial(R.name)]"
+	reagents.add_reagent(R, volume)
+	var/datum/component/heirloom/H = GetComponent(/datum/component/heirloom)
+	desc = H ? "The [H.family_name] family's long-cherished wish is to open this bottle and get its chemical outside. Can you make that wish come true?" : "[desc] [initial(R.name)]."
```
Only humans have family names. It should have another desc for non-human species.
I think I would try this
```
var/speciesdesc = ishuman(H) ? "The [H.family_name] family's" : "Your family's"
desc = H ? "[speciesdesc] long-cherished..."
```

EvilDragonfiend commented on code/modules/reagents/reagent_containers/chem_heirloom.dm:
```diff
@@ -0,0 +1,46 @@
+		..()
+	return
+
+/obj/item/reagent_containers/glass/chem_heirloom/proc/unlock()
```
```
if(!locked)
    return
```

DrDuckedGoose replied:
I have no idea what this does but, I trust you with every fibre of my body.

EvilDragonfiend replied:
my explanation wasn't clear and he misunderstood this. he will fix it later.

DrDuckedGoose committed:
Batch 2

Lizard people and other nerds have weird last names. Fixes that.

Also move naming & desc to late init, to work with components.

DrDuckedGoose committed:
Update negative.dm

DrDuckedGoose commented:
Push is incorrect, name and desc is moved to component called proc

DrDuckedGoose committed:
Batch 3

Review Changes.

PowerfulBacon reviewed (changes requested):
There are definitely other ways to take chemicals out of this than just from attack procs.
The best way to do it would be to fake the contents of the bottle until the end of the round, and then injecting the chemicals at the end.
So it contains nothing, but the sprite shows stuff and then at the end inject a bunch of chems.

EvilDragonfiend commented:
> There are definitely other ways to take chemicals out of this than just from attack procs. The best way to do it would be to fake the contents of the bottle until the end of the round, and then injecting the chemicals at the end. So it contains nothing, but the sprite shows stuff and then at the end inject a bunch of chems.

hmm, if me, if the bottle hasn't the volume of when it was generated, I would send an admin notice that someone opened the bottle.
Give them a good compliment, then check their log and fix it.

DrDuckedGoose committed:
Batch 4

Fix interp & moves chems to populate at roundend.

DrDuckedGoose reviewed (commented):
Chems populate at round end, which should be fine.
I'm not sure what Dragon meant by > Admins can still call the proc. That should be fine?

DrDuckedGoose committed:
Batch 5

- Fix issue with naming relying on heirloom call. Admins can now spawn the bottle in without issues.
- Remove fill icon threshhold, the sprite doesn't need/support it.

DrDuckedGoose committed:
Batch 6

Review changes PT.2

PowerfulBacon commented on code/modules/reagents/reagent_containers/chem_heirloom.dm:
```diff
@@ -0,0 +1,48 @@
+/obj/item/reagent_containers/glass/chem_heirloom/Initialize(mapload, vol)
+	..()
+	update_icon()
+	roundend_callback = CALLBACK(src, .proc/unlock)
+	SSticker.OnRoundend(roundend_callback)
```
Needs to be removed on destroy

PowerfulBacon commented on code/modules/reagents/reagent_containers/chem_heirloom.dm:
```diff
@@ -0,0 +1,48 @@
+//Chemist's heirloom
+
+/obj/item/reagent_containers/glass/chem_heirloom
+	volume = 100
```
I think the volume should be 0 until the round end, so you can't put stuff into this either

PowerfulBacon commented on code/modules/reagents/reagent_containers/chem_heirloom.dm:
```diff
@@ -0,0 +1,48 @@
+/obj/item/reagent_containers/glass/chem_heirloom/attackby(obj/item/I, mob/user, params)
+	if(!locked)
+		..()
+	return
```
This would be redundant if the volume was 0 I believe, may require some testing.
Also returns at the end are pointless since the end of the proc returns anyway

DrDuckedGoose replied:
Yeah, that works fine. I didn't find any immediate issues.

DrDuckedGoose committed:
Batch 7

Review changes
- Changes attack proc early returns in favor of volume adjustment

EvilDragonfiend commented:
hmm, I believe the point was that you can examine what's inside.
How about adding a fake examine text? there are cases examining chems
* Normal examine without science goggles
* Normal examine with science goggles
* Touch the bottle with chem examine cartridge

Once unlocked, the fake description should be removed too.

DrDuckedGoose committed:
Batch 8

- Fix obscure volume init bug

PowerfulBacon commented on code/modules/reagents/reagent_containers/chem_heirloom.dm:
```diff
@@ -24,25 +25,21 @@
+
+/obj/item/reagent_containers/glass/chem_heirloom/Destroy()
+	. = ..()
+	qdel(roundend_callback)
```
You would have to remove it from the roundend callback list you added it to when you called OnRoundend unless the subsystem registers for qdel already.

DrDuckedGoose replied:
I suppose it would but, better safe than sorry.

PowerfulBacon commented on code/modules/reagents/reagent_containers/chem_heirloom.dm:
```diff
@@ -24,25 +25,21 @@
 	locked = FALSE
 	spillable = TRUE
 	reagent_flags = OPENCONTAINER
+	say("Add [rand_cont] [volume]")
```
debug message

DrDuckedGoose committed:
Batch 9

- Remove debug message
- Name & desc update on open to match the state
- Address various examine cases
- Properly remove roundend callbacks
- Address chem-cartridge interaction

DrDuckedGoose committed:
Batch 9.1

- Fix double examine
     Thanks Dragon, for being so attentive

DrDuckedGoose committed:
Update chem_heirloom.dm

DrDuckedGoose committed:
Update chem_heirloom.dm

DrDuckedGoose committed:
better content implementation - 3 6 22

DrDuckedGoose committed:
try some resolves - 17 6 22

DrDuckedGoose committed:
Merge https://github.com/BeeStation/BeeStation-Hornet into annoying-thing
Merge conflicts :)

DrDuckedGoose committed:
test

DrDuckedGoose committed:
test

DrDuckedGoose committed:
Update chemical.dmi

DrDuckedGoose committed:
Update chemical.dmi

DrDuckedGoose commented:
My god man, fucking git
